types/index.d.ts

```diff
@@ -31,52 +34,100 @@ interface AssembleOptions {
 export function discoverProtocol(transport: Service.Transport, callback: Callback<any>): void;
 export function extractFileHeader(filePath: string, options?: any): void;
-export function parse(schemaOrProtocolIdl: string, options?: any): Service.Protocol | Type.Type; // TODO
-export function readProtocol(protocolIdl: string, options?: Partial<ReaderOptions>): Service.Protocol;
+export function infer(args: any): any;
+export function parse(schemaOrProtocolIdl: string, options?: any): Protocol | Type; // TODO
+export function readProtocol(protocolIdl: string, options?: Partial<ReaderOptions>): Protocol;
 export function readSchema(schemaIdl: string, options?: Partial<ReaderOptions>): Schema;
 // TODO streams
-// TODO types
-
-export namespace Type {
-  interface Type {
-    // TODO clone(val, opts)
-    // TODO compare
-    // TODO compareBuffers(buf1, buf2)
-    // TODO createResolver(type, opts)
-    // TODO decode(buf, pos, resolver)
-    // TODO encode(val, buf, pos)
-    // TODO equals(type)
-    // TODO fingerprint(algorithm)
-    fromBuffer(buffer: Buffer, resolver: any, noCheck: boolean): Type; // TODO
-    // TODO fromString(str)
-    // TODO inspect()
-    // TODO isValid(val, opts)
-    // TODO random()
-    // TODO schema(opts)
-    toBuffer(value: object): Buffer;
-    // TODO toJSON()
-    // TODO toString(val)
-    // TODO wrap(val)
-  }
-
-  function forSchema(schema: Schema): Type;
-  function forValue(value: object): Type;
-  // TODO function forTypes(types, opts)
-  function isType(arg: any): boolean; // TODO
+
+// TODO more specific types than `any`
+export class Type {
+  constructor(schema: any, opts: any);
+  clone(val: any, opts?: any): any;
+  compare(val1: any, val2: any): number;
+  compareBuffers(buf1: any, buf2: any): number;
+  createResolver(type: any, opts?: any): any;
+  decode(buf: any, pos?: any, resolver?: any): any;
+  encode(val: any, buf: any, pos?: any): any;
+  equals(type: any): any;
+  fingerprint(algorithm?: any): any;
+  fromString(str: any): any;
+  readonly aliases: any;
+  readonly name: string|undefined;
+  readonly branchName: string|undefined;
+  readonly typeName: string;
+  inspect(): string;
+  isValid(val: any, opts?: any): any;
+  random(): Type;
+  schema(opts?: any): any;
+  toJSON(): string;
+  toString(val?: any): any;
+  wrap(val: any): any;
+  fromBuffer(buffer: Buffer, resolver: any, noCheck: boolean): Type; // TODO
+  toBuffer(value: object): Buffer;
+  static forSchema(schema: Schema, opts?: any): Type;
+  static forTypes(types: any, opts?: any): Type;
+  static forValue(value: object, opts?: any): Type;
+  static isType(arg: any): boolean;  // TODO remaining args
 }
 
-export namespace Service {
-  interface Protocol {
-    [key: string]: any; // TODO object should be further specified
-  }
+// TODO: Can this project remove Protocol from types completely, it's a deprecated export?
```

Comment:
Yes, I think it'd be best to remove the protocol class entirely from here.

Reply:
class Protocol is still referenced by other parts of this file. The references aren't marked as deprecated.